bayes_kit/drghmc.py

```diff
@@ -0,0 +1,437 @@
+        probability and was rejected, propose another sample; if a proposal has high
+        acceptance probablity but was rejected by chance, do not propose another
+        sample. More details found in section 3.2 of Modi et al.
+
+        A heuristic proposal probability that achieve this is eqn. 31 in Modi et al:
+        the probability of making proposal k+1, after rejecting proposal k, is the
+        probability of rejecting proposal k. The proposal probability of making the
+        first proposal, however, is always 1.
+
+        This proposal probability modifies the proposal kernel, which alters detailed
+        balance. Thus the probability of accepting a proposal is multiplied by the
+        hasting term of the propopsed sample divided by the hastings term of the
+        current sample.
+
+        Args:
+            hastings: log probability of rejecting all previous proposals
+
+        Returns:
+            float: log probability of making another proposal
+        """
+        return hastings if self._probabilistic else 0.0
+
+    def sample(self) -> DrawAndLogP:
+        """Sample from target distribution with DrGhmcDiag sampler.
+
+        From current sample (theta, rho), propose new sample (theta_prop, rho_prop)
+        and compute its acceptance probability. If accepted, return the new sample and
+        its log density; if rejected, propose a new sample and repeat. Stop when the
+        maximum number of proposals is reached or when probabilistic delayed rejection
+        terminates early.
+
+        Flip momemntum after each proposal to ensure detailed balance. Flip momentum
+        before return statement for momentum persistence in Generalized HMC.
+
+        Returns:
+            DrawAndLogP: Tuple of (sample, log density)
+        """
+        self._rho = self._rng.normal(
+            loc=self._rho * np.sqrt(1 - self._dampening),
+            scale=self._dampening,
+            size=self._dim,
+        )
+        logp_cur = self.joint_logp(self._theta, self._rho)
+        hastings_cur, reject_logp = 0.0, 0.0
+
+        for k in range(self._num_proposals):
+            pdr = self._prob_of_delayed_rejection(reject_logp)
+            if not np.log(self._rng.uniform()) < pdr:
+                break
+
+            stepsize, steps = self._stepsize_list[k], self._steps_list[k]
+            theta_prop, rho_prop = self.leapfrog(
+                self._theta, self._rho, stepsize, steps
+            )
+            rho_prop = -rho_prop
+
+            accept_logp, logp_prop = self.accept(
+                theta_prop, rho_prop, k, hastings_cur, logp_cur
+            )
+            if np.log(self._rng.uniform()) < accept_logp:
+                self._theta, self._rho = theta_prop, rho_prop
+                logp_cur = logp_prop
+                break
+
+            reject_logp = np.log1p(-np.exp(accept_logp))
+            hastings_cur += reject_logp
+        self._rho = -self._rho  # always flip momentum in GHMC
```

Comment:
Don't we only flip the momentum on an accept?  On a reject, we should be right back where we started, which means not flipping.

Reply:
I meant to write that we negate the momentum in GHMC. Thus the momentum is *flipped* only upon an accept but is *negated* unconditionally. 

I was looking at this helpful line from the [MEADS paper](https://proceedings.mlr.press/v151/hoffman22a/hoffman22a.pdf): "if we accepted the leapfrog move, this will undo the negation from the previous step, but if we rejected the leapfrog move then it will make the chain “bounce” and reverse course."